docs/rfcs/017-asset-modified-cleanup.md

```diff
@@ -0,0 +1,127 @@
+
+## Brief
+
+As part of the delivery channels work, it's possible to modify an asset to either change a delivery channel, or update said delivery channel and leave orphaned assets in S3.  This RFC is to discuss how to remove these assets safely and the various permutations of this removal logic.
+
+## API
+
+Currently, when an asset is modified, a message is published to an SNS topic by the `AssetNotificationSender` with the `before` and `after` of the asset.  This process can then fan out to SQS queue which is then handled by a service that cleans up modified assets in a similar process to how `DELETE` requests are handled.
+
+### Changes
+
+As part of this process there will need to be some changes made to the API, which are as follows:
+
+- AssetModifiedMessage needs to be raised whenever an asset is changed. It's currently only happening for single image requests - PUT, POST or PATCH `/customers/{c}/spaces/{s}/images/{i}` as well as `/reingest` too.
+  - Needs to happen for bulk operations (batch PATCH + queue).
+  - There's already logic to handle batch sending of notifications, just need to get list of messages to send.
+- It's possible to send `AssetModified` messages that aren't required to be ingested (such as metadata changes). In order to reduce churn, an attribute should be added to the request that indicates the asset will be ingested by engine
+  - This attribute should be called something like `EngineNotified`
+- asset requires the `DeliveryChannelPolicyId` to work out differences in policies (this should be there already)
+- asset requires `roles` as changes to roles can mean the `info.json` needs to be removed
+- API should not be responsible for deciding how cleanup is conducted
+
+## AWS
+
+As part of the changes, there will need to be some changes to the AWS estate to support the changes to asset modified.  Primarily, this will be updating the current SNS topic and adding an SQS listener to listen for asset modified notifications.  Additionally, the queue needs to be set to only listen for assets that have a `EngineNotified` attribute. 
+
+### Specifics
+
+- `delay_seconds` will be used to delay messages being added to the queue, to avoid issues with multiple ingests
+- a dead letter queue should be added, as well as `visiblityTimeout` and `maxRetries`, this will help to avoid issues where assets are continually reprocessed
+
+## Handler
+
+After being added to the SQS queue, it needs to be handled by the cleanup handler being extended.  The reason to use a .Net application, is that there will be a lot of shared logic with Engine (for example, working out storage keys) and that access methods for the database and S3 have been written for the DLCS. Additionally, a new implementation of `ImessageHandler` will be created that's designed to handle `AssetModified`
+
+### Specifics
+
+- Assets should be checked for ingestion complete from the `Finished` property before being processed as well as the `Error` property

```

Comment:
might be best with the `Before` value - mainly thinking this as it's more relevant to the system than a timestgamp on a message.

For `Ingesting` I don't think it would hurt - so will add in